### PR TITLE
Strip out premium content blocks from og description tag

### LINKED ISSUE #18998: Premium Content: Subscriber content is accessible in some sources and social media previews.
Reported by @jmdodd in pbMlHh-HM-p2.

The subscriber content of a Premium Content block can be viewed if inspecting some HTML sources (see the linked P2). Posts shared to social media, such as Facebook, might also include subscriber content.

### PR DESCRIPTION
**Pull request by stacimc**
Fixes #18998

#### Changes proposed in this Pull Request:

A short excerpt of post content is added to the og:description Open Graph meta tag. If the post has premium content blocks, the paid subscriber content can be 'leaked' through the tags.

#18654 resolved this by applying `do_blocks` to the description but caused issues with Newspack blocks. This PR is an alternative approach. It removes the description tag entirely if the post contains premium content, unless the author has explicitly set an `excerpt` on the post.

##### Test setup:
* Apply this patch to your sandbox.
* Create a new post.
* View the post on the frontend and open the inspector. Search for the og:description tag.

##### Test cases:

- [ ] For a post that does not contain any premium content blocks, the `og:description` tag exists and contains an excerpt of the post content.
- [ ] For a post that does not contain any premium content but has an `excerpt` set in the post settings, the `og:description` tag exists and is set to the excerpt.
- [ ] For a post containing one or more premium content blocks and no `excerpt`, the `og:description` tag _does not exist_.
- [ ] For a post containing one or more premium content blocks _and_ an `excerpt` set in the post settings, the `og:description` tag exists and is set to the excerpt.
- [ ] For a password-protected post regardless of content and whether an `excerpt` is set, the `og:description` tag dos not exist.

##### Screenshots

_Description tag:_

Screen Shot 2021-03-17 at 11 00 02 AM

### CODE REVIEW

**Review by andrewserong — DISMISSED** [review]
Thanks for updating this @stacimc! This is testing well for me:

✅ For a post that does not contain any premium content blocks, the `og:description` tag exists and contains an excerpt of the post content.
✅  For a post that does not contain any premium content but has an `excerpt` set in the post settings, the `og:description` tag exists and is set to the excerpt.
✅  For a post containing one or more premium content blocks and no `excerpt`, the `og:description` tag does not exist.
✅  For a post containing one or more premium content blocks and an `excerpt` set in the post settings, the `og:description` tag exists and is set to the excerpt.
✅  For a password-protected post regardless of content and whether an `excerpt` is set, the `og:description` tag dos not exist.

I just left an optional comment about using the `has_block` function to check for the `premium-content/container` block instead of `strpos`, but it's totally optional since that function winds up performing essentially the same check anyway.

File: projects/plugins/jetpack/functions.opengraph.php [hunk 1]
```diff
@@ -130,12 +130,14 @@ function jetpack_og_tags() {
...
+			 * Do not use the post content if it contains premium content.
 			 */
-			$excerpt = ! empty( $data->post_excerpt )
-				? $data->post_excerpt
-				: explode( '<!--more-->', $data->post_content )[0];
-
-			$tags['og:description'] = jetpack_og_get_description( $excerpt );
+			if ( ! empty( $data->post_excerpt ) ) {
+				$tags['og:description'] = jetpack_og_get_description( $data->post_excerpt );
+			} elseif ( strpos( $data->post_content, 'wp:premium-content/container' ) === false ) {
```
[the remark this answers is not in the document]
Comment (stacimc) [reply projects/plugins/jetpack/functions.opengraph.php]: Done! Didn't realize I was reinventing the wheel :) I actually strongly prefer `has_block` for readability as well, thanks!

**Review by apeatling — APPROVED** [review]
Tested and all works well for me. :+1:

File: projects/plugins/jetpack/functions.opengraph.php (hunk 1 above)
[the remark this answers is not in the document]
Comment (andrewserong) [reply projects/plugins/jetpack/functions.opengraph.php]: Nice! Thanks for updating it 🙂

**Review by andrewserong — APPROVED** [review]
LGTM, too!

**Review by jeherve — APPROVED** [review]
LGTM. It does mean that we won't have any description for posts with the block though; should we aim (in a follow-up PR maybe) to remove the whole block from the post and build a new OG Description tag from the filtered content?

It also won't solve the issue when folks use other plugins to handle OG tags, such as Yoast SEO.

**Comment by stacimc** [comment]
> It does mean that we won't have any description for posts with the block though; should we aim (in a follow-up PR maybe) to remove the whole block from the post

I actually initially implemented this to use `preg_replace` to strip out the premium content block and build the description from the remaining content (you can see that as the first commit on this PR). I changed it to remove the tag entirely because that's what we do when a post is password protected and it seemed analogous -- but perhaps the original approach makes more sense.

**Comment by stacimc** [comment]
Merging this and creating  258-gh-Automattic/view-design to follow up.

**Comment by stacimc** [comment]
r223719

**Pull request merged**